### PR TITLE
Android implementation using ARCore

### PR DESCRIPTION
**Pull request by yulieth9109**

### CODE REVIEW

File: android/src/main/kotlin/com/example/pocketape/ARCoreManager.kt [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        arSceneView.scene.addOnUpdateListener {
+            arSceneView.arFrame?.let { handleFrameUpdate(it) }
+        }
+    }
+
+    private fun handleFrameUpdate(frame: Frame) {
+        val camera = frame.camera
+        val pose = camera.displayOrientedPose
+        currentPosition = pose.translation
+        sendMeasure()
```
Comment (Robbendebiene) [inline android/src/main/kotlin/com/example/pocketape/ARCoreManager.kt]: Do not store position in `currentPosition` directly pass it to `sendMeasure(pose.translation)` and remove `private var currentPosition: FloatArray? = null`

Should look like this at the end:
```kotlin
    private fun handleFrameUpdate(frame: Frame) {
        sendMeasure(frame.camera.displayOrientedPose.translation)
       // or even simpler directly call plugin.sendEventToFlutter(frame.camera.displayOrientedPose.translation)
    }
```

Same on iOS with `var currentPosition: SIMD4<Float>!`. General programmer rule: store as less state as possible.

File: android/src/main/kotlin/com/example/pocketape/ARCoreManager.kt [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    fun startSession() {
+        try {
+            session?.resume()
+            arSceneView.resume()
```
Comment (Robbendebiene) [inline android/src/main/kotlin/com/example/pocketape/ARCoreManager.kt]: See if there is a restart() method or so that makes sure it starts freshly.

File: android/src/main/kotlin/com/example/pocketape/ARCoreManager.kt [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+    init {
+        if (ContextCompat.checkSelfPermission(context, android.Manifest.permission.CAMERA)
+            == PackageManager.PERMISSION_GRANTED) {
+            initializeARSession()
+            initializeARSceneView()
+        }
+        else {
+            Toast.makeText(context, "Camera permission is required for AR functionality.", Toast.LENGTH_LONG).show()
+        }
```
Comment (Robbendebiene) [inline android/src/main/kotlin/com/example/pocketape/ARCoreManager.kt]: Try to request permissions in android, preferably when the user calls "startMeassure"
Also we have to make sure that the ARScene is not running while the user hasn't pressed "startMeasure" yet.

File: android/src/main/kotlin/com/example/pocketape/PocketapePlugin.kt [hunk 4]
```diff
@@ -3,33 +3,73 @@ package com.example.pocketape
...
-    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "pocketape")
-    channel.setMethodCallHandler(this)
-  }
-
-  override fun onMethodCall(call: MethodCall, result: Result) {
-    if (call.method == "getPlatformVersion") {
-      result.success("Android ${android.os.Build.VERSION.RELEASE}")
-    } else {
-      result.notImplemented()
+class PocketapePlugin : FlutterPlugin, MethodCallHandler {
```
Comment (Robbendebiene) [inline android/src/main/kotlin/com/example/pocketape/PocketapePlugin.kt]: Try adding `StreamHandler` to the extension list.

Then you can get rid of `private var eventHandler: MyStreamHandler? = null` and simply pass `eventChannel.setStreamHandler(this)`

In case this can be applied on iOS as well, please do it.

File: android/src/main/kotlin/com/example/pocketape/PocketapePlugin.kt [hunk 5]
```diff
@@ -3,33 +3,73 @@ package com.example.pocketape
...
+                ARCoreManager.stopSession()
+                result.success(null)
+            }
+            else -> result.notImplemented()
+        }
+    }
+
+    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding)  {
+        channel.setMethodCallHandler(null)
+        eventChannel.setStreamHandler(null)
```
Comment (Robbendebiene) [inline android/src/main/kotlin/com/example/pocketape/PocketapePlugin.kt]: make sure we stop and cleanup the AR session

File: example/android/app/src/main/kotlin/com/example/pocketape/pocketape_example/MainActivity.kt [hunk 6]
```diff
@@ -0,0 +1,5 @@
+package com.example.pocketape.pocketape_example
+
+import io.flutter.embedding.android.FlutterActivity
+
+class MainActivity: FlutterActivity()
```
Comment (Robbendebiene) [inline example/android/app/src/main/kotlin/com/example/pocketape/pocketape_example/MainActivity.kt]: Please figure out if this file has to be committed to github.

**Review by Robbendebiene — CHANGES_REQUESTED** [review]
on the iOS side you can probably remove ` var lastFrame: ARFrame!`

File: android/src/main/kotlin/com/example/pocketape/ARCoreManager.kt (hunk 1 above)
Comment (yulieth9109) [reply android/src/main/kotlin/com/example/pocketape/ARCoreManager.kt]: Done

File: android/src/main/kotlin/com/example/pocketape/ARCoreManager.kt (hunk 2 above)
Comment (yulieth9109) [reply android/src/main/kotlin/com/example/pocketape/ARCoreManager.kt]: There is no restart method, we will have to destroy the session and create it again every time we start (I think that will make the start point slow), However, I checked why I always begin from coordinates 0.0.0, it is because I changed the scene completely and it looks it start again.  I shared a video in the chat

File: android/src/main/kotlin/com/example/pocketape/ARCoreManager.kt (hunk 3 above)
Comment (yulieth9109) [reply android/src/main/kotlin/com/example/pocketape/ARCoreManager.kt]: It is not possible to do it from the plugin since its context is not an Activity, I tried and It does not work. But we can move the configuration of the session in the Start method and it that case, as I mentioned before, it will slow down the start of the measurement.

Additionally, the session will not run if we do not start it (with the resume() method)

File: android/src/main/kotlin/com/example/pocketape/PocketapePlugin.kt (hunk 4 above)
Comment (yulieth9109) [reply android/src/main/kotlin/com/example/pocketape/PocketapePlugin.kt]: Done

File: example/android/app/src/main/kotlin/com/example/pocketape/pocketape_example/MainActivity.kt (hunk 6 above)
Comment (yulieth9109) [reply example/android/app/src/main/kotlin/com/example/pocketape/pocketape_example/MainActivity.kt]: I realize now that I mistakenly created another organization inside example/android, which is why it was created automatically. However, that file is part of the setup for a Flutter project. You can find it here: [MainActivity.kt](https://github.com/OPENER-next/pocketape/blob/main/example/android/app/src/main/kotlin/com/example/pocketape_example/MainActivity.kt). 

Now I understand why it bothers you

File: example/lib/main.dart [hunk 7]
```diff
@@ -12,6 +12,8 @@ void main() {
 class MyApp extends StatelessWidget {
   const MyApp({super.key});
 
+
+
```
Comment (Robbendebiene) [inline example/lib/main.dart]: Minor: avoid unnecessary changes

File: ios/Classes/ARKitManager.swift [hunk 8]
```diff
@@ -7,7 +7,6 @@ class ARKitManager: NSObject, ARSessionDelegate, ARSCNViewDelegate {
     
     var trackingStateOK: Bool = false
     var lastFrame: ARFrame!
```
Comment (Robbendebiene) [inline ios/Classes/ARKitManager.swift]: Are these variables needed? At least they are not used in your code but maybe they are required by the implementation?

File: lib/pocketape.dart [hunk 9]
```diff
@@ -23,15 +25,20 @@ abstract class Pocketape {
           onDone: controller.close,
         );
         if (_count == 1) {
+          print("Start");
+          if (Platform.isAndroid && !(await _requestPermission())) {
```
Comment (Robbendebiene) [inline lib/pocketape.dart]: Minor: directly call `await Permission.contacts.request().isGranted` this also checks whether it has already been granted.

File: lib/pocketape.dart [hunk 10]
```diff
@@ -23,15 +25,20 @@ abstract class Pocketape {
...
+            controller.addError(CameraPermissionDeniedException("Camera permission denied"));
```
Comment (Robbendebiene) [inline lib/pocketape.dart]: Minor: as a message use: "Camera permission is required on Android in order to use ARCore."

File: lib/pocketape.dart [hunk 11]
```diff
@@ -23,15 +25,20 @@ abstract class Pocketape {
...
+            return;
+          }
```
Comment (Robbendebiene) [inline lib/pocketape.dart]: Maybe I was wrong.
If we add it here we falsely increment the counter and also subscribe to the stream.
I guess we should add it at the very top of the onListen function so we do not even subscribe to the event stream when no permission is granted. What do you think?

File: lib/pocketape.dart [hunk 12]
```diff
@@ -23,15 +25,20 @@ abstract class Pocketape {
...
           await _platformChannel.invokeMethod('startMeasure');
         }
       },
       onCancel: () async {
         _count--;
-        subscription.cancel();
         if (_count == 0) {
           await _platformChannel.invokeMethod('stopMeasure');
         }
+        await subscription.cancel();
```
Comment (Robbendebiene) [inline lib/pocketape.dart:40]: I suppose this is done to resolve the async coordinates delay (not starting at 0,0)?

I do not really understand why this improves the situation. The only thing I can think of is that there are still events added to the stream from android which are buffered. Thus if we unsubscribe too early they are kept in the buffer and sent on the next subscription (trace() call)

File: android/src/main/kotlin/com/example/pocketape/ARCoreManager.kt [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+                initializeARSceneView()
+                session?.resume()
+                arSceneView?.resume()
+            }
+            else {
+                Toast.makeText(context, "Camera permission is required for AR functionality.", Toast.LENGTH_LONG).show()
+            }
+        } catch (e: Exception) {
+            e.printStackTrace()
+        }
```
Comment (Robbendebiene) [inline android/src/main/kotlin/com/example/pocketape/ARCoreManager.kt]: Now that we do it in flutter only throw an error here. No toast message.

File: android/src/main/kotlin/com/example/pocketape/ARCoreManager.kt [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+            updateListener = Scene.OnUpdateListener {
+                arSceneView?.arFrame?.let { frame -> handleFrameUpdate(frame) }
+            }
+            updateListener?.let { scene.addOnUpdateListener(it) }
+        }
+    }
+
+    private fun handleFrameUpdate(frame: Frame) {
+        sendMeasure(frame)
+    }
```
Comment (Robbendebiene) [inline android/src/main/kotlin/com/example/pocketape/ARCoreManager.kt]: Minor: not required. Directly use `sendMeasure`

File: android/src/main/kotlin/com/example/pocketape/ARCoreManager.kt [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+
+    public fun onDestroy() {
+        arSceneView?.scene?.let { scene ->
+            updateListener?.let { scene.removeOnUpdateListener(it) }
+        }
+        arSceneView?.destroy()
+        session?.close()
+        session = null
+    }
```
Comment (Robbendebiene) [inline android/src/main/kotlin/com/example/pocketape/ARCoreManager.kt]: move this code inside `stopSession` and there you probably do not need the following lines anymore.

```
        session?.pause()
        arSceneView?.pause()
```

**Review by Robbendebiene — CHANGES_REQUESTED** [review]

File: lib/pocketape.dart (hunk 11 above)
Comment (yulieth9109) [reply lib/pocketape.dart]: Yes, I understand your point @Robbendebiene , The point is to check only at the start, and if I move it above I will check every time. Maybe I can include a constructor then it checks when the instance is created?

File: lib/pocketape.dart (hunk 12 above)
Comment (yulieth9109) [reply lib/pocketape.dart:40]: Yes, that's what I think.

File: android/src/main/kotlin/com/example/pocketape/ARCoreManager.kt (hunk 15 above)
Comment (yulieth9109) [reply android/src/main/kotlin/com/example/pocketape/ARCoreManager.kt]: @Robbendebiene Why do we not pause? It is supposed to pause before removing and destroying to guarantee a proper behavior

Comment (Robbendebiene) [reply android/src/main/kotlin/com/example/pocketape/ARCoreManager.kt]: Any docs you can share that point this out? I would assume that detroying/closing does the pause internally.

File: lib/pocketape.dart (hunk 11 above)
Comment (Robbendebiene) [reply lib/pocketape.dart]: I don't necessarily think that the "check everytime" is wrong. The permission can be removed at any time. We could still solve solve this by checking whether count is 0 before requesting the permission as long as it is done on the very top of the function.

File: android/src/main/kotlin/com/example/pocketape/ARCoreManager.kt (hunk 15 above)
Comment (yulieth9109) [reply android/src/main/kotlin/com/example/pocketape/ARCoreManager.kt]: The library documentation mentioned : Releases resources (a significant amount of native heap memory) used by an ARCore session.
Failure to close sessions explicitly may cause your app to run out of native memory and crash. If your app contains a single AR-enabled activity, it is recommended that you call close() from the activity's onDestroy method.

This method will take several seconds to complete. To prevent blocking the main thread, call pause() on the main thread, and then call close() on a background thread.

It is not safe to call methods on this session or other objects obtained from this session while the session is being closed and after the session is closed.
When closing a session that was created with Session. Feature#SHARED_CAMERA, the camera device must be closed via CameraDevice#close() before calling this method.
Creation of a new session on a different thread will not be blocked by the ongoing call to close() on a background thread.

File: lib/pocketape.dart (hunk 9 above)
Comment (yulieth9109) [reply lib/pocketape.dart]: @Robbendebiene _requestPermission() method is not only checking, but it is also checking and if they are not granted ask the user to approve, in that case, that line can not be replaced.

File: lib/pocketape.dart (hunk 11 above)
Comment (yulieth9109) [reply lib/pocketape.dart]: ok, I will move it to the top

File: lib/pocketape.dart (hunk 9 above)
Comment (Robbendebiene) [reply lib/pocketape.dart]: `await Permission.camera.request().isGranted` should do exactly what we need (check and request when needed). Docs say:

> Call request() on a Permission to request it. If it has already been granted before, nothing happens.
request() returns the new status of the Permission.

```
if (await Permission.contacts.request().isGranted) {
  // Either the permission was already granted before or the user just granted it.
}
```

File: android/src/main/kotlin/com/example/pocketape/ARCoreManager.kt (hunk 15 above)
Comment (Robbendebiene) [reply android/src/main/kotlin/com/example/pocketape/ARCoreManager.kt]: So this is interesting and all, but it doesn't mention that we have to call pause before destroy, or am I missing something? They mention something about performance, but we are anyway executing everything in the same thread so this doesn't apply here.

Comment (yulieth9109) [reply android/src/main/kotlin/com/example/pocketape/ARCoreManager.kt]: The ArScene should be pause if not the app crush.
